app/src/main/java/com/hedvig/app/feature/embark/passages/previousinsurer/retrieveprice/StartDataCollectionUseCase.kt

```diff
@@ -0,0 +1,101 @@
+package com.hedvig.app.feature.embark.passages.previousinsurer.retrieveprice
+
+import com.apollographql.apollo.ApolloClient
+import com.hedvig.android.owldroid.graphql.ExternalInsuranceProviderV2Query
+import com.hedvig.android.owldroid.graphql.InitiateDataCollectionNOMutation
+import com.hedvig.android.owldroid.graphql.InitiateDataCollectionSEMutation
+import com.hedvig.app.feature.settings.Market
+import com.hedvig.app.feature.settings.MarketManager
+import com.hedvig.app.util.apollo.QueryResult
+import com.hedvig.app.util.apollo.safeQuery
+import java.lang.IllegalArgumentException
+import java.util.UUID
+
+interface StartDataCollectionUseCase {
+    suspend fun startDataCollection(
+        personalNumber: String,
+        insuranceProvider: String,
+    ): DataCollectionResult
+}
+
+class StartDataCollectionUseCaseImpl(
+    val apolloClient: ApolloClient,
+    val marketManager: MarketManager
+) : StartDataCollectionUseCase {
+
+    override suspend fun startDataCollection(
+        personalNumber: String,
+        insuranceProvider: String,
+    ): DataCollectionResult {
+        val reference = UUID.randomUUID().toString()
+
+        val mutation = createMutation(
+            reference = reference,
+            personalNumber = personalNumber,
+            insuranceProvider = insuranceProvider
+        )
+
+        return when (val result = apolloClient.mutate(mutation).safeQuery()) {
+            is QueryResult.Success -> getCollectionStatus(reference)
+            is QueryResult.Error -> DataCollectionResult.Error.NetworkError(result.message)
+        }
+    }
+
+    private suspend fun getCollectionStatus(
+        reference: String
+    ): DataCollectionResult {
+        val query = ExternalInsuranceProviderV2Query(reference)
+        return when (val result = apolloClient.query(query).safeQuery()) {
+            is QueryResult.Success -> {
+                val extraInfo = result.data?.externalInsuranceProvider?.dataCollectionStatusV2?.extraInformation
+                extraInfo
+                    ?.asSwedishBankIdExtraInfo
+                    ?.autoStartToken
+                    ?.let { DataCollectionResult.Success.SwedishBankId(it) } ?: extraInfo
+                    ?.asNorwegianBankIdExtraInfo
+                    ?.norwegianBankIdWords
+                    ?.let { DataCollectionResult.Success.NorwegianBankId(it) }
+                    ?: DataCollectionResult.Error.NoData
+            }
+            is QueryResult.Error -> DataCollectionResult.Error.QueryError
+        }
+    }
+
+    private fun createMutation(
+        reference: String,
+        personalNumber: String,
+        insuranceProvider: String
+    ) = when (marketManager.market) {
```

Comment:
With the function returning such different types, maybe set the return type explicitly? Looking at this in GitHub I don't understand what's the common class of the NO/SEMutation is.

Reply:
The type is actually `Mutation<out Operation.Data, out Operation.Data, Operation.Variables>`, that feels almost more confusing. The function name also gives a hint that this will return a mutation.